Add trusted application functionality

CounterPillow opened the pull request:
This lets users apply for trusted status, given certain minimum requirements. Moderators can then review the applications, giving a recommendation, and administrators can accept or reject them.

If an application is accepted or rejected, the user receives an e-mail about it.

Markdown images are not rendered in applications to prevent browsers from sending automatic requests to untrusted webservers.

Users who have had their application rejected cannot re-apply for a set amount of days.

TheAMM commented on nyaa/models.py:
```diff
@@ -847,6 +867,77 @@ def is_rangebanned(cls, ip):
+
+
+class TrustedApplication(db.Model):
+    __tablename__ = 'trustedapps'
```
`__tablename__ = 'trusted_applications'`

TheAMM commented on nyaa/views/account.py:
```diff
@@ -223,6 +223,48 @@ def profile():
+        if trusted_form.validate():
+            ta = models.TrustedApplication()
+            ta.submitter_id = flask.g.user.id
+            ta.why_want = trusted_form.why_want_trusted.data
```
I'd `.strip()` these two just because, although I forget if WTF passes the minimum length even if it's spaces.

CounterPillow replied:
Do we get any issue if we don't strip them? In the very synthetic case of stripping them, it could break code blocks at the start of the field opened with 4 spaces. Obviously not something to worry about in this context though.

TheAMM replied:
`r.strip()`, then. It's simply a small useless abuse vector, pasting 4k spaces. 4k of data is not much to worry about, but hey, it's user input, and it's clean to `.rstrip()` it.

TheAMM commented on nyaa/views/account.py:
```diff
@@ -223,6 +223,48 @@ def profile():
+            deny_reasons.append('You already are trusted.')
+        if not flask.g.user.satisfies_trusted_reqs:
+            deny_reasons.append('You do not satisfy the minimum requirements.')
+        if (models.TrustedApplication.query.
```
`.first()` is fine instead of `.count() > 0`, although if you chose this for performance, :woman_shrugging:.

CounterPillow commented on nyaa/models.py:
```diff
@@ -643,6 +644,25 @@ def created_utc_timestamp(self):
+                filter(t_flavor.flags.op('&')(int(TorrentFlags.REMAKE)).is_(False)).scalar()
+            num_total += uploads or 0
+            downloads_total += dls or 0
+        print("Uploads: {}, DLs: {}".format(num_total, downloads_total))
```
oops I should probably get rid of this

CounterPillow commented:
Do you think I should index `submitter_id` in `TrustedApplication`? We do filter by it quite a lot, though I do not know if this is relevant enough.

Hamuko commented on nyaa/views/account.py:
```diff
@@ -223,6 +223,48 @@ def profile():
+    deny_reasons = []
+    if flask.request.method == 'POST':
+        trusted_form = forms.TrustedForm(flask.request.form)
+        if trusted_form.validate():
```
`deny_reasons` is never computed/checked for POST requests, so if you don't pass minimum requirements, you can just directly POST the data.

CounterPillow replied:
They would have a hard time getting past trusted_form.validate() without a csrf token which they never get because they never get the form, but sure, I've changed to code so that it checks the `deny_reasons` both paths.

CounterPillow commented:
I've addressed AMM's comments and rebased the migration on top of @nyaadev's. If someone reviews this and finds it to be satisfactory, then I think this is ready to merge.

Arylide commented:
Wow
